awx/main/scheduler/task_manager_models.py

```diff
@@ -139,3 +228,20 @@ def get_instance_groups_from_task_cache(self, task):
             logger.warn(f"No instance groups in cache exist, defaulting to global instance groups for task {task}")
             return task.global_instance_groups
         return igs
+
+
+class TaskManagerModels:
+    def __init__(self, **kwargs):
+        # We want to avoid calls to settings over and over in loops, so cache this information here
+        kwargs['control_task_impact'] = kwargs.get('control_task_impact', settings.AWX_CONTROL_NODE_TASK_IMPACT)
+        kwargs['controlplane_ig_name'] = kwargs.get('controlplane_ig_name', settings.DEFAULT_CONTROL_PLANE_QUEUE_NAME)
+        self.instances = TaskManagerInstances(**kwargs)
+        self.instance_groups = TaskManagerInstanceGroups(task_manager_instances=self.instances, **kwargs)
+        if kwargs.get('init_consumed_capacity', False) or 'tasks' in kwargs:
+            tasks = kwargs.get('tasks', None)
+            if tasks is None:
+                from awx.main.models import UnifiedJob
+
+                tasks = UnifiedJob.objects.filter(status__in=['running', 'waiting']).only('task_impact', 'controller_node', 'execution_node', 'instance_group')
+            for task in tasks:
+                self.instance_groups.consume_capacity(task)
```

Comment:
I would like to better isolate `init_consumed_capacity` from the logic that runs in the task manager by a separate `@classmethod` constructor.

like

```python
@classmethod
def init_with_consumed_capacity(cls, **kwargs):
    tmm = cls(**kwargs)
    tasks = kwargs.get('tasks', None)
    if tasks is None:
        from awx.main.models import UnifiedJob

        tasks = UnifiedJob.objects.filter(status__in=['running', 'waiting']).only('task_impact', 'controller_node', 'execution_node', 'instance_group')
    for task in tasks:
        self.instance_groups.consume_capacity(task)
```

Then call by `TaskManagerModels.init_with_consumed_capacity()`

Reply:
sounds good

Reply:
done